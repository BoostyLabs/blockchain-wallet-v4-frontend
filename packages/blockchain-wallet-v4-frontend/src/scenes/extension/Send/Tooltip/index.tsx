import React from 'react'
import { IconAlert } from '@blockchain-com/icons'
import styled from 'styled-components'

import { Text } from 'blockchain-info-components'

const TooltipWrapper = styled.div`
  position: absolute;
  top: 90px;
  right: 270px;
  border-radius: 8px;
  padding: 20px 40px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 135px;
  border: 1px solid #ffa133;
  color: white;
  background: none;
`

const Tooltip: React.FC = () => {
  return (
    <TooltipWrapper id='invalidAddress'>
      <IconAlert height='24px' color='#FFA133' width='24px' />
      <Text size='14px' lineHeight='21px' weight={600} color='white'>
        Invalid address
      </Text>
    </TooltipWrapper>
  )
}

export default Tooltip
