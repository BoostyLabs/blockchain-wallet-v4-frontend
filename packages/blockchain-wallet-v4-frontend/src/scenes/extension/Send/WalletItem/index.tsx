import React from 'react'
import { IconBlockchainCircle } from '@blockchain-com/icons'
import styled from 'styled-components'

import { Text } from 'blockchain-info-components'

const WalletWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: flex-start;
  color: white;
  cursor: pointer;
`

const AccountInfo = styled.div`
  margin-left: 16px;
  display: flex;
  flex-direction: column;
`

const WalletItem: React.FC<{ walletAddress: string }> = ({ walletAddress }) => {
  return (
    <WalletWrapper id='walletItem'>
      <IconBlockchainCircle height='32px' color='#473BCB' width='32px' />
      <AccountInfo>
        <Text size='12px' lineHeight='18px' style={{ color: '#98A1B2' }} weight={500}>
          Blockchain.com trading account
        </Text>
        <Text size='14px' lineHeight='21px' color='white' weight={500}>
          {walletAddress}
        </Text>
      </AccountInfo>
    </WalletWrapper>
  )
}

export default WalletItem
