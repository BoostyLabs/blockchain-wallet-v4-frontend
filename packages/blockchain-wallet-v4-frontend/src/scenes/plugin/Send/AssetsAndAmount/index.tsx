import React, { useState } from 'react'
import { FormattedMessage } from 'react-intl'
import { IconCheckCircle, IconChevronDownV2, IconCloseCircle } from '@blockchain-com/icons'
import Amount from 'blockchain-wallet-v4-frontend/src/scenes/plugin/Send/AssetsAndAmount/Amount'
import SelectCrypto from 'blockchain-wallet-v4-frontend/src/scenes/plugin/Send/AssetsAndAmount/Assets'
import CryptoItem from 'blockchain-wallet-v4-frontend/src/scenes/plugin/Send/AssetsAndAmount/Assets/Asset'
import styled from 'styled-components'

import { Button, Text } from 'blockchain-info-components'
import { Flex } from 'components/Flex'

import { AvailableSteps } from '..'

const Wrapper = styled(Flex)`
  flex-direction: column;
  justify-content: space-between;
  position: relative;
  width: 100%;
  height: 100%;
`

const SavedWallet = styled(Flex)`
  align-items: center;
  padding: 12px 16px;
  margin: 17px 0;
  width: 248px;
  overflow-wrap: break-word;
  word-break: break-word;
  border-radius: 8px;
  background: ${(props) => props.theme.exchangeLogin};
  color: ${(props) => props.theme.grey500};
`

const SavedWalletWrapper = styled(Flex)`
  align-items: center;
  justify-content: space-between;
  width: 100%;
`

const AssetWrapper = styled(Flex)`
  justify-content: space-between;
  align-items: center;
  margin: 5px auto;
  padding: 4px 0px;
  width: 100%;
  border-radius: 8px;
  background: black;
  cursor: pointer;
  &:hover {
    background: ${(props) => props.theme.exchangeLogin};
  }
`

const Label = styled(Text)`
  font-size: 12px;
  line-height: 18px;
  font-weight: 500;
  color: ${(props) => props.theme.grey400};
  margin-top: 35px;
`

type AssetsAndAmountProps = {
  address: string
  changeStep: (step: AvailableSteps) => void
}

const AssetsAndAmount: React.FC<AssetsAndAmountProps> = ({ address, changeStep }) => {
  const [isSelectCryptoListVisible, setIsSelectCryptoListVisible] = useState<boolean>(false)

  const [selectedCoin, setSelectedCoin] = useState<string>('ETH')

  // changes select crypto list visibility
  const changeSelectCryptoListVisibility = () => {
    setIsSelectCryptoListVisible(!isSelectCryptoListVisible)
  }

  const selectCoin = (coin: string) => {
    setSelectedCoin(coin)
    setIsSelectCryptoListVisible(!isSelectCryptoListVisible)
  }

  return (
    <Wrapper>
      <SavedWalletWrapper>
        <IconCloseCircle
          height='16px'
          width='16px'
          onClick={() => changeStep(AvailableSteps.Recents)}
        />
        <SavedWallet>
          <Text size='14px' lineHeight='21px' weight={500}>
            {address}
          </Text>
          <IconCheckCircle width='24px' height='24px' color='white' />
        </SavedWallet>
      </SavedWalletWrapper>
      <Label>
        <FormattedMessage id='plugin.send.assets_label' defaultMessage='Assets' />
      </Label>
      <AssetWrapper>
        <CryptoItem coin={selectedCoin} selectCoin={selectCoin} />
        <IconChevronDownV2 color='grey400' />
      </AssetWrapper>
      <Label>
        <FormattedMessage id='plugin.send.amount_label' defaultMessage='Amount' />
      </Label>
      <Amount coin={selectedCoin} />
      <Flex justifyContent='space-between'>
        <Button nature='white-transparent' data-e2e='pluginSendCancelButton' height='48px'>
          <FormattedMessage id='plugin.send.cancel_button' defaultMessage='Cancel' />
        </Button>
        <Button height='48px' data-e2e='pluginSendNextButton'>
          <FormattedMessage id='plugin.send.next_button' defaultMessage='Next' />
        </Button>
      </Flex>
      {isSelectCryptoListVisible && (
        <SelectCrypto
          changeSelectCryptoListVisibility={changeSelectCryptoListVisibility}
          selectCoin={selectCoin}
        />
      )}
    </Wrapper>
  )
}

export default AssetsAndAmount
