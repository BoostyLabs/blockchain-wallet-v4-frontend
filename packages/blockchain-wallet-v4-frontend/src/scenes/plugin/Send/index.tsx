import React, { useEffect, useState } from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import { actions } from 'data'

import SelectAddress from './SelectAddress'

// defines available steps to send amount
export enum AvailableSteps {
  SelectAddress,
  FirstStep,
  SecondStep
}

// TODO: finalize with erc20 tokens list
export enum AvailableCoins {
  ETH = 'ETH',
  USDC = 'USDC',
  USDT = 'USDT'
}

const Send = (props) => {
  const [step, setStep] = useState<number>(AvailableSteps.SelectAddress)
  const [coin, setCoin] = useState<string>(AvailableCoins.ETH)

  const { sendActions } = props
  // indicates if recents wallets are shown
  const isRecentsWalletsShown: boolean = step === AvailableSteps.SelectAddress

  // changes step
  const changeStep = (step: AvailableSteps) => {
    setStep(step)
  }

  useEffect(() => {
    sendActions.initialized(coin)
  }, [coin])

  return (
    <>
      {isRecentsWalletsShown && (
        <SelectAddress changeStep={changeStep} step={step} coin={coin} {...props} />
      )}
    </>
  )
}

const mapDispatchToProps = (dispatch) => ({
  sendActions: bindActionCreators(actions.components.sendEth, dispatch)
})

export default connect(null, mapDispatchToProps)(Send)
