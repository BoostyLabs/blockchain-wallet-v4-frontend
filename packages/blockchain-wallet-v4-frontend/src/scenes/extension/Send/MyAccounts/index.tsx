import React from 'react'

import { WalletsWrapper } from 'blockchain-wallet-v4-frontend/src/scenes/extension/Send'
import WalletItem from 'blockchain-wallet-v4-frontend/src/scenes/extension/Send/WalletItem'

const MyAccounts: React.FC = () => {
  // TODO: mock accounts wallets data.
  const myAccounts: string[] = [
    'A4Pd7dB08D75679bbE90b7B1eDfB8BE3a16897Ee77',
    'B4Ad7dB08D75679bbE90b7B1eDfB8BE3a16897Ee77'
  ]

  return (
    <WalletsWrapper>
      {myAccounts.map((address: string) => (
        <WalletItem walletAddress={address} key={address} />
      ))}
    </WalletsWrapper>
  )
}

export default MyAccounts
