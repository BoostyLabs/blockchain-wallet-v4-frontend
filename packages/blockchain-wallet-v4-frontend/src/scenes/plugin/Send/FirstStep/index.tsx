import React from 'react'
import { connect, ConnectedProps } from 'react-redux'
import getData from 'blockchain-wallet-v4-frontend/src/modals/Eth/SendEth/FirstStep/selectors'
import { SEND_FORM } from 'blockchain-wallet-v4-frontend/src/modals/SendCrypto/model'
import { AvailableSteps } from 'blockchain-wallet-v4-frontend/src/scenes/plugin/Send'
import { bindActionCreators, compose } from 'redux'
import { reduxForm } from 'redux-form'

import { BlockchainLoader } from 'blockchain-info-components'
import { actions } from 'data'

import FirstStepSuccess from './template.success'

const FirstStep: React.FC<Props> = (props) => {
  const { changeCoin, coin, data, history, sendActions } = props

  const firstStepConfirm = () => {
    sendActions.sendEthFirstStepSubmitClicked()
  }

  // Re-inits payment provider for selected coin, i.e 'ETH' or erc20 token
  const reInitCoinPayment = () => {
    sendActions.initialized(coin)
  }

  return data.cata({
    Failure: (message) => <>{message}</>,
    Loading: () => <BlockchainLoader height='24px' width='24px' />,
    NotAsked: () => null,
    Success: (value) => {
      return (
        <FirstStepSuccess
          changeCoin={changeCoin}
          coin={coin}
          firstStepConfirm={firstStepConfirm}
          history={history}
          reInitCoinPayment={reInitCoinPayment}
          {...value}
        />
      )
    }
  })
}

const mapDispatchToProps = (dispatch) => ({
  sendActions: bindActionCreators(actions.components.sendEth, dispatch)
})

const mapStateToProps = (state, props) => ({
  data: getData(state, props.coin)
})

const connector = connect(mapStateToProps, mapDispatchToProps)

type FirstStepProps = {
  changeCoin: (coin: string) => void
  coin: string
  history: {
    push: (path: string) => void
  }
}

type Props = FirstStepProps & ConnectedProps<typeof connector>

const enhance = compose(
  reduxForm<{}, Props>({
    destroyOnUnmount: false,
    form: SEND_FORM
  }),
  connector
)

export default enhance(FirstStep) as React.ComponentType<FirstStepProps>
