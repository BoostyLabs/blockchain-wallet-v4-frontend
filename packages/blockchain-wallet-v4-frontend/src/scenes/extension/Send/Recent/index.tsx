import React from 'react'
import { WalletsWrapper } from 'blockchain-wallet-v4-frontend/src/scenes/extension/Send'
import WalletItem from 'blockchain-wallet-v4-frontend/src/scenes/extension/Send/WalletItem'

const Recent: React.FC = () => {
  // TODO: mock recent wallets data.
  const recentWallets: string[] = [
    'J4Pd7dB08D75679bbE90b7B1eDfB8BE3a16897Ee77',
    'J4Ad7dB08D75679bbE90b7B1eDfB8BE3a16897Ee77'
  ]

  return (
    <WalletsWrapper>
      {recentWallets.map((address: string) => (
        <WalletItem walletAddress={address} key={address} />
      ))}
    </WalletsWrapper>
  )
}

export default Recent
