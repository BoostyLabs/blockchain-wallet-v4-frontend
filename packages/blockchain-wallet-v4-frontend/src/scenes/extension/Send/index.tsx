import React, { useState } from 'react'
import { Link, Route, Switch } from 'react-router-dom'
import { IconClose } from '@blockchain-com/icons'
import MyAccounts from 'blockchain-wallet-v4-frontend/src/scenes/extension/Send/MyAccounts'
import Recent from 'blockchain-wallet-v4-frontend/src/scenes/extension/Send/Recent'
import Tooltip from 'blockchain-wallet-v4-frontend/src/scenes/extension/Send/Tooltip'
import styled from 'styled-components'

import { Text } from 'blockchain-info-components'

const Wrapper = styled.div`
  padding: 27px 24px;
  height: 100%;
  background: #0e121b;
`

const IconCloseWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: flex-end;
  cursor: pointer;
`

const InputWrapper = styled.div`
  position: relative;
  display: flex;
`

export const WalletsWrapper = styled.div`
  margin-top: 36px;
  display: flex;
  align-items: flex-start;
  flex-direction: column;
`

const Input = styled.input`
  margin: 17px auto;
  padding: 16px 12px;
  width: 100%;
  border-radius: 8px;
  border: 1px solid #619ff7;
  font-weight: 500;
  font-size: 16px;
  line-height: 24px;
  color: #677184;
  background: none;
  outline: none;
`

const Tabs = styled.div`
  margin-top: 155px;
  display: flex;
  align-items: center;
  justify-content: center;
  color: #98a1b2;
  .active {
    background: #98a1b2;
    color: black;
  }
`

const Tab = styled.button`
  padding: 3px 10px;
  height: 27px;
  font-size: 14px;
  line-height: 22px;
  color: white;
  border-radius: 8px;
  background: none;
  border: none;
  outline: none;
  cursor: pointer;
`

enum SendTabs {
  myAccounts = 'My accounts',
  recent = 'Recent'
}

const Send = () => {
  const [activeTab, setActiveTab] = useState<SendTabs | null>(null)
  const [walletAddress, setWalletAddress] = useState<string>('')
  const [isTooltipVisible, setIsTooltipVisible] = useState(false)

  // changes and validates wallet address
  const changeWalletAddress = (e: React.ChangeEvent<HTMLInputElement>) => {
    const walletPattern = /[a-z0-9]{42}/i
    setWalletAddress(e.target.value)
    if (!walletPattern.test(e.target.value)) {
      setIsTooltipVisible(true)
      return
    }
    setIsTooltipVisible(false)
    window.location.replace('/#/extension/sending')
  }

  // changes active tab.
  const changeActiveTab = (activeTab: SendTabs) => {
    setActiveTab(activeTab)
  }

  // Returns active tab className
  const getActiveTabClassName: (tab: string) => string = (tab: string) =>
    tab === activeTab ? 'active' : ''

  return (
    <Wrapper>
      <IconCloseWrapper>
        <IconClose height='24px' width='24px' color='white' cursor='pointer' />
      </IconCloseWrapper>
      <Text size='20px' lineHeight='30px' color='white' weight={500}>
        Send to
      </Text>
      <InputWrapper>
        <Input
          placeholder='|Search, public address (0x), or ENS '
          value={walletAddress}
          onChange={changeWalletAddress}
          id='sendToWalletAddress'
        />
        {isTooltipVisible && <Tooltip message='invalid address' />}
      </InputWrapper>
      <Tabs>
        <Link to='/extension/send/recent'>
          <Tab
            id='recentTab'
            className={getActiveTabClassName(SendTabs.recent)}
            onClick={() => changeActiveTab(SendTabs.recent)}
          >
            {SendTabs.recent}
          </Tab>
        </Link>
        <Link to='/extension/send/my-accounts'>
          <Tab
            id='myAccountsTab'
            className={getActiveTabClassName(SendTabs.myAccounts)}
            onClick={() => changeActiveTab(SendTabs.myAccounts)}
          >
            {SendTabs.myAccounts}
          </Tab>
        </Link>
      </Tabs>
      <Switch>
        <Route path='/extension/send/recent' component={Recent} />
        <Route path='/extension/send/my-accounts' component={MyAccounts} />
      </Switch>
    </Wrapper>
  )
}

export default Send
