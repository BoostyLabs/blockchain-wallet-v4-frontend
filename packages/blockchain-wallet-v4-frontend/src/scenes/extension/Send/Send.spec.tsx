import React from 'react'
import Send from 'blockchain-wallet-v4-frontend/src/scenes/extension/Send'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

describe('Send extension scene', (): void => {
  const container = shallow(<Send />)
  const walletAddress = '0x0d7dB08D75679bbE90b7B1eDfB8BE3a16897Ee17'

  it('Send scene renders correctly', (): void => {
    const tree = toJson(container)
    expect(tree).toMatchSnapshot()
  })

  it('should set the wallet address value on change event', (): void => {
    container.find('#sendToWalletAddress').simulate('change', {
      target: {
        value: walletAddress
      }
    })
    expect(container.find('#sendToWalletAddress').prop('value')).toEqual(walletAddress)
  })

  it('recent tab should have active className on click', (): void => {
    container.find('#recentTab').simulate('click')
    expect(container.find('#recentTab').hasClass('active')).toBeTruthy()
  })

  it('my accounts tab should have active className on click', (): void => {
    container.find('#myAccountsTab').simulate('click')
    expect(container.find('#myAccountsTab').hasClass('active')).toBeTruthy()
  })
})
