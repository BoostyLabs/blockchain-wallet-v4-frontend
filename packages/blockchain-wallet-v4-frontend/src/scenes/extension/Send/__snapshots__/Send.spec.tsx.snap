// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Send extension scene Send scene renders correctly 1`] = `
<styled.div>
  <styled.div>
    <SvgIconClose
      color="white"
      cursor="pointer"
      height="24px"
      width="24px"
    />
  </styled.div>
  <Text
    altFont={false}
    capitalize={false}
    color="white"
    cursor="inherit"
    italic={false}
    lineHeight="30px"
    opacity={1}
    size="20px"
    uppercase={false}
    weight={500}
  >
    Send to
  </Text>
  <styled.div>
    <styled.input
      id="sendToWalletAddress"
      onChange={[Function]}
      placeholder="|Search, public address (0x), or ENS "
      value=""
    />
  </styled.div>
  <styled.div>
    <Link
      replace={false}
      to="/extension/send/recent"
    >
      <styled.button
        className=""
        id="recentTab"
        onClick={[Function]}
      >
        Recent
      </styled.button>
    </Link>
    <Link
      replace={false}
      to="/extension/send/my-accounts"
    >
      <styled.button
        className=""
        id="myAccountsTab"
        onClick={[Function]}
      >
        My accounts
      </styled.button>
    </Link>
  </styled.div>
  <Switch>
    <Route
      component={[Function]}
      path="/extension/send/recent"
    />
    <Route
      component={[Function]}
      path="/extension/send/my-accounts"
    />
  </Switch>
</styled.div>
`;
